### PR TITLE
Add tidb-cluster-on-centos template

### PR DESCRIPTION
**Pull request by onlymellb**
tidb cluster on centos template allows users to define the PD/TiKV/TiDB components and cluster monitoring. More detailed information about TiDB can be found [here](https://github.com/pingcap/docs).

### CODE REVIEW

**Comment by onlymellb** [comment]
@singhkays and @sedouard, CI have test passed, could you please help to check my PR, thanks.

File: tidb-cluster-on-centos/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Deploy a TiDB cluster on CentOS virtual machines
+
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2Ftidb-cluster-on-centos%2Fazuredeploy.json" target="_blank">
+  <img src="http://azuredeploy.net/deploybutton.png"/>
+</a>
+<a href="http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2Ftidb-cluster-on-centos%2Fazuredeploy.json" target="_blank">
+  <img src="http://armviz.io/visualizebutton.png"/>
+</a>
+
+This template deploys a TiDB cluster on the CentOS virtual machines. This template also provisions a storage account, virtual network, network security groups, availability sets, load balancer, public IP addresses and network interfaces required by the installation. The template also creates 1 publicly accessible VM acting as a "monitor server" and "jumpbox" allowing to ssh into the PD/TiKV/TiDB nodes for diagnostics or troubleshooting purposes. In addition, and created a load balancer for TiDB servers so that users can connect to the database through this.
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/README.md]: Typo in the last sentence... "In addition, and..."

File: tidb-cluster-on-centos/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+        "description": "Administrator user name used when provisioning virtual machines"
+      }
+    },
+    "adminPassword": {
+      "type": "securestring",
+      "metadata": {
+        "description": "Administrator password used when provisioning virtual machines"
+      }
+    },
+    "storageAccountNamePrefix": {
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: move this to a variable and uniqueString, or just use managed disks

File: tidb-cluster-on-centos/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+      }
+    },
+    "addressPrefix": {
+      "type": "string",
+      "defaultValue": "10.0.0.0/8",
+      "metadata": {
+        "description": "The network address space for the virtual network"
+      }
+    },
+    "routeTableName": {
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: move to variable

File: tidb-cluster-on-centos/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+      }
+    },
+    "tikvVmSize": {
+      "type": "string",
+      "defaultValue": "Standard_D12_v2",
+      "metadata": {
+        "description": "Virtual machine size for the tikv virtual machine"
+      }
+    },
+    "artifactsLocation": {
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#samples-that-contain-extra-artifacts-custom-scripts-nested-templates-etc) for how to handle artifacts

File: tidb-cluster-on-centos/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+      "offer": "CentOS",
+      "sku": "7.2",
+      "version": "latest"
+    },
+    "pdIpPrefix": "[substring(parameters('pdSubnetPrefix'), 0, lastIndexOf(parameters('pdSubnetPrefix'), '.'))]",
+    "tidbIpPrefix": "[substring(parameters('tidbSubnetPrefix'), 0, lastIndexOf(parameters('tidbSubnetPrefix'), '.'))]",
+    "tikvIpPrefix": "[substring(parameters('tikvSubnetPrefix'), 0, lastIndexOf(parameters('tikvSubnetPrefix'), '.'))]",
+    "monitorIpAddr": "[concat(variables('pdIpPrefix'), '.254')]",
+    "vhdContainerName": "vhd-tidbcluster",
+    "environmentAzureCloud": ".blob.core.windows.net",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: remove these and use reference to get the uri/endpoint for storage

File: tidb-cluster-on-centos/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+    "vmStorageAccountDomain": "[variables(concat('environment', parameters('environment')))]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks', parameters('virtualNetworkName'))]",
+    "templateBaseUrl": "[parameters('artifactsLocation')]",
+    "sharedTemplateUrl": "[concat(variables('templateBaseUrl'), 'nested/shared-resources.json')]",
+    "commonSettings": {
+      "pdAVSName": "pdAvailSet",
+      "tidbAVSName": "tidbAvailSet",
+      "tikvAVSName": "tikvAvailSet"
+    },
+    "pdStorageSettings": {
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: use reference for all of these or use managed disks

File: tidb-cluster-on-centos/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+      "destinationVhdsContainer": "[concat('https://', parameters('storageAccountNamePrefix'), 'tidb', variables('vmStorageAccountDomain'), '/', variables('vhdContainerName'), '/')]"
+    },
+    "tikvStorageSettings": {
+      "vhdStorageAccountName": "[concat(parameters('storageAccountNamePrefix'), 'tikv')]",
+      "destinationVhdsContainer": "[concat('https://', parameters('storageAccountNamePrefix'), 'tikv', variables('vmStorageAccountDomain'), '/', variables('vhdContainerName'), '/')]"
+    },
+    "pdNetworkSettings": {
+      "subnetName": "[parameters('pdSubnetName')]",
+      "subnetPrefix": "[parameters('pdSubnetPrefix')]",
+      "subnetRef": "[concat(variables('vnetID'), '/subnets/', parameters('pdSubnetName'))]"
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: use resourceId instead of concat

File: tidb-cluster-on-centos/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+    },
+    "tidbNetworkSettings": {
+      "subnetName": "[parameters('tidbSubnetName')]",
+      "subnetPrefix": "[parameters('tidbSubnetPrefix')]",
+      "lbName": "tidb-svc",
+      "subnetRef": "[concat(variables('vnetID'), '/subnets/', parameters('tidbSubnetName'))]"
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: use resourceId instead of concat

File: tidb-cluster-on-centos/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+    },
+    "tikvNetworkSettings": {
+      "subnetName": "[parameters('tikvSubnetName')]",
+      "subnetPrefix": "[parameters('tikvSubnetPrefix')]",
+      "subnetRef": "[concat(variables('vnetID'), '/subnets/', parameters('tikvSubnetName'))]"
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: use resourceId instead of concat

File: tidb-cluster-on-centos/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+    },
+    "scriptsToDownload": "[concat(parameters('artifactsLocation'), 'scripts/tidb-cluster-install.sh')]",
+    "singleQuote": "'"
+  },
+  "resources": [
+    {
+      "name": "shared-resources",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: update apiVersion

File: tidb-cluster-on-centos/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+          "addressPrefix": {
+             "value": "[parameters('addressPrefix')]"
+          }
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Resources/deployments",
+      "name": "pd-node-resources",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: update apiVersion

File: tidb-cluster-on-centos/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+          "adminPassword": {
+            "value": "[parameters('adminPassword')]"
+          }
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Resources/deployments",
+      "name": "tikv-node-resources",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: update apiVersion

File: tidb-cluster-on-centos/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+          "adminPassword": {
+            "value": "[parameters('adminPassword')]"
+          }
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Resources/deployments",
+      "name": "tidb-node-resources",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: update apiVersion

File: tidb-cluster-on-centos/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+          "adminPassword": {
+            "value": "[parameters('adminPassword')]"
+          }
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Resources/deployments",
+      "name": "monitor-node-resources",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: update apiVersion

File: tidb-cluster-on-centos/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,467 @@
...
+            }
+          },
+          "adminPassword": {
+            "value": "[parameters('adminPassword')]"
+          }
+        }
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.json]: update apiVersion

File: tidb-cluster-on-centos/azuredeploy.parameters.json [hunk 16]
```diff
@@ -0,0 +1,45 @@
+{
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "adminUsername": {
+        "value": "tidb"
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.parameters.json]: GEN-UNIQUE

File: tidb-cluster-on-centos/azuredeploy.parameters.json [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    "adminPassword": {
+        "value": "GEN-PASSWORD"
+    },
+    "storageAccountNamePrefix": {
+      "value": "GEN-UNIQUE-5"
+    },
+    "virtualNetworkName": {
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/azuredeploy.parameters.json]: for any values that are not "GEN" values, just use defaultValues on the parameters in the template

File: tidb-cluster-on-centos/nested/monitor-node-resource.json [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+  },
+  "variables": {
+    "publicIPAddressType": "Dynamic",
+    "privateIPAddressType": "Static",
+    "publicIPAddressName": "[concat('pubip-', parameters('machineSettings').machineName)]",
+    "networkInterfacesName": "[concat('nic-', parameters('machineSettings').machineName)]"
+  },
+  "resources": [
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/monitor-node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/monitor-node-resource.json [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "publicIPAllocationMethod": "[variables('publicIPAddressType')]",
+        "dnsSettings": {
+          "domainNameLabel": "[concat('monitor', uniqueString(resourceGroup().id))]"
+        }
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/monitor-node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/monitor-node-resource.json [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+          }
+        ],
+        "networkSecurityGroup": {
+          "id": "[resourceId('Microsoft.Network/networkSecurityGroups', 'monitor-secgroup')]"
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Compute/virtualMachines",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/monitor-node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/node-resource.json [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+  },
+  "variables": {
+    "vmStart": 1,
+    "ipBase": 9,
+    "privateIPAddressType": "Static",
+    "networkInterfacesName": "[concat('nic-', parameters('machineSettings').machineNamePrefix)]"
+  },
+  "resources": [
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/node-resource.json [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+                "id": "[parameters('networkSettings').subnetRef]"
+              }
+            }
+          }
+        ]
+      }
+    },
+    {
+      "type": "Microsoft.Compute/virtualMachines",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+      "metadata": {
+        "Description": "The virtual network address prefix"
+      }
+    }
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[concat(parameters('pdStorageSettings').vhdStorageAccountName)]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "accountType": "Standard_LRS"
+      }
+    },
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[concat(parameters('tidbStorageSettings').vhdStorageAccountName)]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "accountType": "Standard_LRS"
+      }
+    },
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[concat(parameters('tikvStorageSettings').vhdStorageAccountName)]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "accountType": "Standard_LRS"
+      }
+    },
+    {
+      "type": "Microsoft.Compute/availabilitySets",
+      "name": "[parameters('commonSettings').pdAVSName]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "PlatformFaultDomainCount": "3",
+        "PlatformUpdateDomainCount": "5"
+      }
+    },
+    {
+      "type": "Microsoft.Compute/availabilitySets",
+      "name": "[parameters('commonSettings').tidbAVSName]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "PlatformFaultDomainCount": "3",
+        "PlatformUpdateDomainCount": "5"
+      }
+    },
+    {
+      "type": "Microsoft.Compute/availabilitySets",
+      "name": "[parameters('commonSettings').tikvAVSName]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 29]
```diff
@@ -0,0 +1,242 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "PlatformFaultDomainCount": "3",
+        "PlatformUpdateDomainCount": "5"
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 30]
```diff
@@ -0,0 +1,242 @@
...
+      "type": "Microsoft.Network/routeTables",
+      "name": "[parameters('routeTableName')]",
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "routes": [
+        ]
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 31]
```diff
@@ -0,0 +1,242 @@
...
+              "routeTable": {
+                "id": "[resourceId('Microsoft.Network/routeTables', parameters('routeTableName'))]"
+              }
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/shared-resources.json [hunk 32]
```diff
@@ -0,0 +1,242 @@
...
+                "access": "Allow",
+                "priority": 101,
+                "direction": "Inbound"
+              }
+            }
+          ]
+       }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/shared-resources.json]: update apiVersion

File: tidb-cluster-on-centos/nested/tidb-node-resource.json [hunk 33]
```diff
@@ -0,0 +1,214 @@
...
+    "lbPublicIPAddressID": "[resourceId('Microsoft.Network/publicIPAddresses', variables('lbPublicIPAddressName'))]",
+    "frontendLbName": "[concat(variables('lbName'), '-frontend')]",
+    "backendAddrPoolName": "[concat(variables('lbName'), '-backend')]",
+    "lbRulesName": "[concat(variables('lbName'), '-rules')]",
+    "lbProbeName": "[concat(variables('lbName'), '-probe')]",
+    "networkInterfacesName": "[concat('nic-', parameters('machineSettings').machineNamePrefix)]"
+  },
+  "resources": [
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/tidb-node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/tidb-node-resource.json [hunk 34]
```diff
@@ -0,0 +1,214 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "publicIPAllocationMethod": "[variables('lbPublicIPAddressType')]",
+        "dnsSettings": {
+          "domainNameLabel": "[concat('loadbalancer', uniqueString(resourceGroup().id))]"
+        }
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/tidb-node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/tidb-node-resource.json [hunk 35]
```diff
@@ -0,0 +1,214 @@
...
+              "port": "4000",
+              "intervalInSeconds": 5,
+              "numberOfProbes": 2
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/tidb-node-resource.json]: update apiVersion

File: tidb-cluster-on-centos/nested/tidb-node-resource.json [hunk 36]
```diff
@@ -0,0 +1,214 @@
...
+          }
+        ],
+        "networkSecurityGroup": {
+          "id": "[resourceId('Microsoft.Network/networkSecurityGroups', 'tidb-secgroup')]"
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Compute/virtualMachines",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline tidb-cluster-on-centos/nested/tidb-node-resource.json]: update apiVersion

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.

**Pull request closed without merging**